### ISSUE TITLE
add method to silence old API deprecation build warnings

### ISSUE THREAD

**Issue by rgommers** [numpy/numpy#11653]
See https://github.com/cython/cython/issues/2498#issuecomment-409455046. If by the 1.16.0 release there is still no support in Cython for the non-deprecated C API, then we should add a way to silence the build warnings. A simple environment variable that, if set, silences the build warning would be easy to implement.

**Comment by charris** [comment]
We should probably just go ahead and do that rather than waiting on Cython. Admittedly, that reduces the pressure to fix things, but that doesn't seem to have had an effect yet.

**Comment by jbrockmendel** [comment]
Woops, just sent an email to the mailing list before seeing the ball is already rolling.  Pls ignore.

**Comment by mattip** [comment]
Cython 0.29 was released, and can now suppress the warnings. Leaving this open until we upgrade our build system to use it.

**Comment by mattip** [comment]
Since PR #12286 the warnings can now be suppressed by setting `-DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION` from the command line or 
```
#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
```
in code. This allows the old API with direct access to `PyArray_Objectfield` structure data members.

Further work on Cython/NumPy integration:
- cython/cython#2640 will allow `NPY_NO_DEPRECATED_API` to be updated to `NPY_1_16_API_VERSION`
- PR #12284 adds our own `numpy.pxd` which can track our API changes independent of cython.

From my reading of the original issue, #12286 is enough to close this.

**Comment by charris** [comment]
That's not right,  `NPY_NO_DEPRECATED_API` needs to be < `NPY_1_7_API_VERSION`, I suggest using 0.

**Comment by mattip** [comment]
Right, thanks for the correction.

**Comment by rgommers** [comment]
> I suppose we could define `NPY_1_6_API_VERSION`.

That would be useful.

**Comment by charris** [comment]
@rgommers My thought also.

**Comment by charris** [comment]
@rgommers Good enough? We could optionally unconditionally remove the warning.

**Comment by charris** [comment]
Closing this. If we need better than #12286 another issue can be opened.

**Comment by rgommers** [comment]
> @rgommers Good enough? We could optionally unconditionally remove the warning.

Yep, the current status looks fine to me. Unconditional removal should only be done if we abandon the whole thing, and there's another issue for discussing that.
